tests/system/libraries/systemTestSpy.py

```diff
@@ -0,0 +1,207 @@
+# A part of NonVisual Desktop Access (NVDA)
+# Copyright (C) 2018 NV Access Limited
+# This file may be used under the terms of the GNU General Public License, version 2 or later.
+# For more details see: https://www.gnu.org/licenses/gpl-2.0.html
+
+"""This file provides spy and robot library behaviour for NVDA system tests.
+It is copied into the (system test specific) NVDA profile directory. It becomes the '__init__.py' file as part of a
+package. This allows us to share utility methods between the global plugin and the nvdaRobotLib library.
+"""
+import globalPluginHandler
+import threading
+from systemTestUtils import _blockUntilConditionMet
+from logHandler import log
+from time import clock as _timer
+
+import sys
+import os
+log.debug("before pathmod: {}".format(sys.path))
+# Get the path to the top of the package
+TOP_DIR = os.path.abspath(os.path.dirname(__file__))
+# imports that require libraries not distributed with an install of NVDA
+sys.path.append( os.path.join(TOP_DIR, "libs"))
+log.debug("after pathmod: {}".format(sys.path))
+from robotremoteserver import RobotRemoteServer
+
+whitespaceMinusSlashN = '\t\x0b\x0c\r '
+
+
+class SystemTestSpy(object):
+	SPEECH_HAS_FINISHED_SECONDS = 0.5
+
+	def __init__(self):
+		self._nvdaSpeech = [
+			[""],  # initialise with an empty string, this allows for access via [-1]. This is equiv to no speech.
+		]
+		self._allSpeechStartIndex = 0
+		self._speechOccurred = False
+		self.isNvdaStartupComplete = False
+		self.lastSpeechTime = _timer()
+		self._registerWithExtensionPoints()
+
+	def _registerWithExtensionPoints(self):
+		from core import postNvdaStartup
+		from speech import preSpeech
+		postNvdaStartup.register(self._onNvdaStartupComplete)
+		preSpeech.register(self._onNvdaSpeech)
+
+	# callbacks for extension points
+	def _onNvdaStartupComplete(self):
+		self.isNvdaStartupComplete = True
+
+	def _onNvdaSpeech(self, speechSequence=None):
+		if not speechSequence: return
+		with threading.Lock():
+			self._speechOccurred = True
+			self.lastSpeechTime = _timer()
+			self._nvdaSpeech.append(speechSequence)
+
+	# Private helper methods
+	def _flattenCommandsSeparatingWithNewline(self, commandArray):
+		f = [c for commands in commandArray for newlineJoined in [commands, [u"\n"]] for c in newlineJoined]
+		return f
+
+	def _getJoinedBaseStringsFromCommands(self, speechCommandArray):
+		wsChars = whitespaceMinusSlashN
+		baseStrings = [c.strip(wsChars) for c in speechCommandArray if isinstance(c, basestring)]
+		return ''.join(baseStrings).strip()

```

Comment:
Why does this need restripped?

Reply:
It possibly doesn't need to be there. There is the possibility of a trailing/leading newline character, we want to preserve newline characters within the string. But not at the ends, since they are harder to notice, and can make defining the expected text a frustrating experience.